### DISCUSSION TITLE
`refreshAfterWrite` not allowed to nullify non-expired keys?

### DISCUSSION THREAD

**Discussion by SHxKM** [ben-manes/caffeine#1808] [Q&A] [answered]
Requirement:

1. Attempt to refresh an entry `t1` after it has been added
2. Expire the entry `t2` after it has been added, unless it has been "refreshed", in which case its expiry should renew

`refreshAfterWrite` seems to be the ideal way to do it, however, I found something about it that works perfectly for my use-case but was surprising to me: if the `refresh` callback returns null, it will not remove the entry (again, perfect for my use-case, but kinda unintuitive. Here's a test case:

**Update:** it seems like the behavior is different in real environments as opposed to unit-tests, meaning that if `refreshAfterWrite` is due, and the refresh function returns `null`, then the value is nullified and evicted from cache.

```java
@Test
void testRefreshAfterWriteWhenNullIsFetched() throws InterruptedException {
    Cache<String, String>  testCache = Caffeine.newBuilder()
            .refreshAfterWrite(Duration.ofSeconds(1))
            .expireAfterWrite(Duration.ofSeconds(10))
            .build(this::fetchFromRemote);

    testCache.put("foo", "bar");
    assertThat(testCache.getIfPresent("foo")).isEqualTo("bar");

    Thread.sleep(3000);
    assertThat(testCache.getIfPresent("foo")).isEqualTo("bar");

    Thread.sleep(9000);
    assertThat(testCache.getIfPresent("foo")).isNull();
}

private @Nullable String fetchFromRemote(String s) {
      return null;
}
```

Is this because according to caffeine, `refresh` is only for "hydrating" the value, and not ever for removing it?

**Comment by ben-manes** [comment, marked as answer]
Sorry, I'm unsure what the question is. A refresh to a null value should remove the entry. The various loads try to follow their Map compute counterpart, here a refresh is an asynchronous `computeIfPresent`. If you implement the `CacheLoader` explicitly then you can override `reload` or `asyncReload` to leverage the old value. The `asyncReload` is the callback and you can delay the future, e.g. to batch reloads or simply never complete one. If the entry is updated, removed, or evicted then the in-flight reload will be abandoned as the cache tries to have linearizable semantics for reasoning against (or close enough to be a good mental model). I'm not sure if any of that helped but let me know either way.

**Comment by SHxKM** [comment]
> Sorry, I'm unsure what the question is. **A refresh to a null value should remove the entry**. The various loads try to follow their Map compute counterpart, here a refresh is an asynchronous `computeIfPresent`.

That what the question indeed, in my unit tests I saw the value was not nullified, but I guess that's because the background thread didn't get a chance to run. Thanks @ben-manes!
